src/s3/putobject.cpp

```diff
@@ -88,36 +124,217 @@ asio::awaitable<void> irods::s3::actions::handle_putobject(
         }
 
         uint64_t read_buffer_size = irods::s3::get_put_object_buffer_size_in_bytes();
+        uint64_t max_read_buffer_size = irods::s3::get_put_object_max_buffer_size_in_bytes();
         std::cout << "read buffer size = " << read_buffer_size << std::endl;
+        std::cout << "max read buffer size = " << max_read_buffer_size << std::endl;
         std::vector<char> buf_vector(read_buffer_size);
         parser.get().body().data = buf_vector.data();
         parser.get().body().size = read_buffer_size; 
 
         response.set("Etag", path.c_str());
         response.set("Connection", "close");
 
-        while (!parser.is_done()) {
-            parser.get().body().data = buf_vector.data();
-            parser.get().body().size = read_buffer_size;
-            //            auto read = co_await beast::http::async_read_some(socket, buffer, parser,
-            //            asio::use_awaitable);
-            auto read = beast::http::read_some(socket, buffer, parser);
-            size_t read_bytes = read_buffer_size - parser.get().body().size;
+        if (special_chunked_header) {
+            std::cout << "read buffer size = " << read_buffer_size << std::endl;
 
-            try {
-                d.write((char*) buf_vector.data(), read_bytes);
-            }
-            catch (std::exception& e) {
-                std::cout << e.what() << std::endl;
-            }
 
-            if (ec == beast::http::error::need_buffer) {
-                ec = {};
+            boost::beast::error_code ec;
+
+            enum class parsing_state {
+                HEADER_BEGIN, HEADER_CONTINUE, BODY, END_OF_CHUNK, DONE, ERROR
+            };
+
+            parsing_state current_state = parsing_state::HEADER_BEGIN;
+            std::string the_bytes;
+
+            size_t chunk_size = 0;
+
+            // While we have more data to be read from the socker into the_bytes or
+            // we haven't exhausted the_bytes.
+            while (!parser.is_done() || the_bytes != "") {
+
+                // check to see if we are done parsing the chunks
+                if (current_state == parsing_state::DONE || current_state == parsing_state::ERROR) {
+                    break;
+                }
+
+                parser.get().body().data = buf_vector.data();
+                parser.get().body().size = read_buffer_size;
+                
+                if (!parser.is_done()) {
+                    parser.eager(true);
+                    auto read = beast::http::read_some(socket, buffer, parser, ec);
+
+                    if(ec == beast::http::error::need_buffer) {
+                        read_buffer_size *= 2;
```

Comment:
this is the most important part - add a comment about doubling...

Reply:
done